src/Microsoft.AspNetCore.Mvc.Razor/RazorView.cs

```diff
@@ -99,24 +99,32 @@ public string Path
             }
 
             _bufferScope = context.HttpContext.RequestServices.GetRequiredService<IViewBufferScope>();
-            var bodyWriter = await RenderPageAsync(RazorPage, context, invokeViewStarts: true);
+            var bodyWriter = await RenderPageContentAsync(RazorPage, context, invokeViewStarts: true);
             await RenderLayoutAsync(context, bodyWriter);
         }
 
-        private async Task<RazorTextWriter> RenderPageAsync(
+        private async Task<ViewBufferTextWriter> RenderPageContentAsync(
             IRazorPage page,
             ViewContext context,
             bool invokeViewStarts)
         {
-            Debug.Assert(_bufferScope != null);
-            var buffer = new ViewBuffer(_bufferScope, page.Path);
-            var razorTextWriter = new RazorTextWriter(context.Writer, buffer, _htmlEncoder);
+            var writer = context.Writer as ViewBufferTextWriter;
+            if (writer == null)
+            {
+                Debug.Assert(_bufferScope != null);
+
+                // If we get here, this is likely the top-level page (not a partial) - this means
+                // that context.Writer is wrapping the output stream. We need to buffer, so create a buffered writer.
+                var buffer = new ViewBuffer(_bufferScope, page.Path, pageSize: 256);
+                writer = new ViewBufferTextWriter(buffer, context.Writer.Encoding, _htmlEncoder, context.Writer);
+            }
```

Comment:
Another big bugfix here. If we were rendering a partial, we'd wrap a buffered writer in a buffer. This way we can reuse instead of being greedy.